src/next/useSearchParams.ts

```diff
@@ -1,63 +1,66 @@
-import { useCallback, useEffect, useMemo, useRef } from "react"
-import {
-  usePathname,
-  useSearchParams as useNextSearchParams,
-  useRouter
-} from "next/navigation"
+import { useCallback, useEffect, useRef } from "react"
+import { useSearchParams as useNextSearchParams } from "next/navigation"
 
 type SearchParamsSetterValue =
   | URLSearchParams
   | ((prevSearchParams: URLSearchParams) => URLSearchParams)
 
 type SetSearchParams = (newSearchParams: SearchParamsSetterValue) => void
 
-const useSearchParams = (): [URLSearchParams, SetSearchParams] => {
-  const pathname = usePathname()
-  const router = useRouter()
-  const search = useNextSearchParams()
-
+/**
+ * Returns a function for setting search params in the browser URL.
+ *  - the first call within a single render cycle will add a new entry in the
+ *   history stack via pushState
+ *  - subsequent calls will replace the previous entry in the history stack
+ *    and will receive the previously updated value in setter function
+ *
+ * NOTE: This is intended for use with NextJS, but could work with any framework
+ * that supports direct usage of window.history.pushState / replaceState.
+ */
+const useSetSearchParams = (): SetSearchParams => {
   /**
    * Keep track of whether navigate has been called in the current render cycle
    * to avoid adding extra entries in the history stack.
    */
   const hasNavigatedRef = useRef(false)
-  const searchParams = useMemo(() => new URLSearchParams(search), [search])
-  /**
-   * Keep track of the current searchParams value so that updater functions can
-   * use the current value rather than value from previous render.
-   */
-  const searchParamsRef = useRef(searchParams)
 
   useEffect(() => {
     hasNavigatedRef.current = false
-    /**
-     * Each render, sync the ref with the current state value.
-     * This is necessary in case search params has changed via some source other
-     * than this hook (e.g., browser navigation).
-     */
-    searchParamsRef.current = searchParams
   })
 
-  const setSearchParams: SetSearchParams = useCallback(
-    nextValue => {
-      const newParams =
-        typeof nextValue === "function" ?
-          nextValue(searchParamsRef.current) :
-          nextValue
-
-      searchParamsRef.current = newParams
-
-      if (hasNavigatedRef.current) {
-        router.replace(`${pathname}?${newParams}`, { scroll: false })
-      } else {
-        router.push(`${pathname}?${newParams}`, { scroll: false })
-      }
-
-      hasNavigatedRef.current = true
-    },
-    [pathname, router]
-  )
-  return [searchParams, setSearchParams]
+  const setSearchParams: SetSearchParams = useCallback(nextValue => {
+    const newParams =
+      typeof nextValue === "function" ?
+        nextValue(new URLSearchParams(window.location.search)) :
+        nextValue
+
+    if (hasNavigatedRef.current) {
```

Comment:
I'm interested in how this _has-run_ mechanism works in relation to the render cycle. If called each parent render cycle - and if navigate is the result of user action (checking a facet filter) I'd expect a render cycle to have been triggered before the user can navigate again (it does pre-date this PR though).